rex-release/out.sh

```diff
@@ -55,21 +55,70 @@ upload-release() {
   # books files with explicit content type set because they don't have extensions, loaded unversioned so no cloudfront caching
   aws s3 sync --exclude 'service-worker.js' --content-type 'text/html' --cache-control 'max-age=0' "$path/books/" "s3://$bucket/rex/releases/$version/books"
 
-  # configure redirects
-  while read -r row; do
-    from=$(jq -r '.from' <<< "$row")
-    to=$(jq -r '.to' <<< "$row")
+  # upload redirects in parallel, limited by MAX_S3_CONCURRENCY (default 10)
+  MAX_S3_CONCURRENCY=${MAX_S3_CONCURRENCY:-10}
+  s3_pids=()
+
+  # track redirects we've created
+  created_from_redirects=()
+
+  # wait for the first pid in s3_pids and remove it from the array
+  wait_pop_s3_pid() {
+    wait "${s3_pids[0]}" || {
+      echo "one of the s3 put-object commands failed"
+      exit 1
+    }
+    s3_pids=("${s3_pids[@]:1}")
+  }
+
+  # process_redirect: $1=version $2=from $3=to $4=skip_to_check (true/false)
+  process_redirect() {
+    version="$1"
+    from="$2"
+    to="$3"
+    skip_to_check="$4"
 
     from_exists=$(release-file-exists "$version" "$from")
-    to_exists=$(release-file-exists "$version" "${to%$q*}")
 
-    if [ -n "$from_exists" ] || { [[ "$to" == /books* ]] && [ -z "$to_exists" ]; }; then
-      echo "cannot create redirection from $from to $to, aborting"
-      exit 1;
+    if [ -n "$from_exists" ] || {
+      [ "$skip_to_check" != "true" ] &&
+      [ -z "$(release-file-exists "$version" "${to%"$q"*}")" ]
```

Comment:
I decided to put the release-file-exists command in the if statement that way we skip it completely in the 2 cases we don't care about $to_exists